### ISSUE TITLE
[RFC] [Intel GPU] Extending TorchFT to Support Intel GPU with XCCL Backend

### ISSUE THREAD

**Issue by siju-samuel** [meta-pytorch/torchft#257]
## Abstract
We propose extending TorchFT to support [Intel XPU](https://docs.pytorch.org/docs/stable/notes/get_start_xpu.html) devices by adapting ProcessGroups to use the "xccl" backend. This extension will enable Intel GPU users to benefit from the same per-step fault tolerance guarantees currently available on CUDA, including checkpointing, elastic recovery, node failure detection, and AFT-style recovery.

## Motivation
TorchFT today provides per-step fault tolerance for PyTorch training, allowing jobs to continue without restarting the entire loop when failures occur. It currently supports CUDA backends (NCCL/Gloo) for FT-DDP, FSDP/HSDP, and checkpointing.

[Intel’s XPU](https://docs.pytorch.org/docs/stable/notes/get_start_xpu.html) (Supported via `torch.xpu` since PyTorch 2.5) is gaining adoption in large-scale training and inference. To fully leverage TorchFT on Intel GPUs, we need native XPU integration with the "xccl" backend. Without this, Intel GPU users cannot take advantage of TorchFT’s core features (fault-tolerant DDP/FSDP, elastic scaling, checkpointing).

Adding XPU support ensures parity with CUDA workflows and expands TorchFT’s utility to a broader hardware ecosystem. The architecture diagram below illustrates TorchFT’s high-level design: (Added XCCL to the Original Diagram)

<img width="1663" height="460" alt="Image" src="https://github.com/user-attachments/assets/ee165aac-c86a-4ef1-98f0-be9a38d8418b" />

### API and Code Changes

1. **Device Checks**
   - Replace CUDA-specific checks with device-agnostic APIs where possible.
   - Example:
     ```python
     # Instead of current CUDA-only
     torch.cuda.is_available()
     torch.cuda.set_device(id)

     # Use device-agnostic APIs for uniform detection across CUDA and XPU
     torch.accelerator.is_available()
     torch.accelerator.set_device(id)
     ```

2. **Streams & Events**
   - Replace `torch.cuda.Stream` / `torch.cuda.Event` with device-agnostic `torch.Stream` / `torch.Event`.

3. **Process Groups**
   - Extend TorchFT’s ProcessGroup wrappers to support:
     - `ProcessGroupXCCL`
     - `ProcessGroupBabyXCCL`

4. **DDP/FSDP/HSDP Wrappers for Tests**
   - Ensure TorchFT’s DDP/HSDP wrappers accept `device_ids=["xpu"]` and support model/input migration to XPU (`model.to("xpu")`).
   - Verify compatibility with FSDP on XPU.

5. **Checkpointing**
   - TorchFT checkpointing layers are device-agnostic. Verify seamless support for XPU as well.
   - Validate save/restore workflows on XPU.

## Feature Parity: CUDA vs XPU

| TorchFT Feature           | CUDA (NCCL/Gloo) | XPU (XCCL) – Proposed Support |
|---------------------------|------------------|-------------------------------|
| **ProcessGroup**          | ✅ Supported     | 🚧 Planned (ProcessGroupXCCL / ProcessGroupBabyXCCL) |
| **Manager**               | ✅ Supported     | 🚧 Planned (XPU Stream/Event handling in manager) |
| **Checkpointing**         | ✅ Supported     | 🚧 Planned (validate XPU for save/restore checkpointing) |
| **DDP/Dtensor/FSDP**      | ✅ Supported     | 🚧 Planned (validate tests with relevant modifications) |
| **Multi-node Failure Recovery** | ✅ Supported | 🚧 Planned (validate XCCL across multi-node clusters) |
| **CI Test Coverage**      | ✅ Supported     | 🚧 Planned |

---

✅ = Supported today  
🚧 = Planned / requires implementation & validation

#### **1. Finalize and Merge Initial PR**
  - [x] Submit the [first PR](https://github.com/pytorch/torchft/pull/260) with:
    - Manager function
    - ProcessGroup integration
    - XPU device handling
    - Stream & event management
  - [x ] Ensure CUDA functionality is unaffected
  
  #### **2. Expand Testing Coverage**
  - [ ] Validate entire TorchFT test suite on single-node XPU
  - [ ] Enable multi-node tests for:
    - `diloco_regression_test.py`
    - `local_sgd_integ_test.py`
  
  #### **3. Performance Validation & Benchmarking**
  - [ ] Benchmark core collectives (`allreduce`, `allgather`, `broadcast`, `reduce_scatter`)
  - [ ] Run standard training workloads:
    - ResNet-50
    - GPT-style models
  - [ ] Scaling efficiency
  
  #### **5. Documentation & User Guide**
  - [ ] Update **README** with XPU usage instructions

## Useful Links
[Getting Started on Intel GPU](https://docs.pytorch.org/docs/stable/notes/get_start_xpu.html) 
[PyTorch XPU Support](https://github.com/intel/torch-xpu-ops)

cc @tushar00jain @jeromean @gujinghui @zhangxiaoli73 @rbabukv

**Comment by gujinghui** [comment]
@zhangxiaoli73 can you review this? Thanks.

**Comment by zhangxiaoli73** [comment]
@siju-samuel XCCL is already announced as one of PyTorch 2.8 release feature https://github.com/pytorch/pytorch/releases#highlights.  The RFC looks good to me. Could you add a next-step checklist so others can see your planned actions and progress?

**Comment by d4l3k** [comment]
Great to see this! Also exciting to see the code already :)

If you guys want to have a meeting/chat about this face to face happy to set one up. Feel free to message me on the PyTorch Slack if you're interested and I can help. cc @mreso @HamidShojanazeri
